llpc/lower/llpcSpirvLowerConstImmediateStore.cpp

```diff
@@ -51,164 +52,180 @@ namespace Llpc {
-        auto storeInst = findSingleStore(allocaInst);
-        if (storeInst && isa<Constant>(storeInst->getValueOperand())) {
-          // Got an aggregate "alloca" with a single store to the whole type.
-          // Do the optimization.
-          convertAllocaToReadOnlyGlobal(storeInst);
-        }
-      }
+  bool changed = false;
+  SmallVector<AllocaInst *> candidates;
+  for (auto &inst : func->getEntryBlock()) {
+    if (auto allocaInst = dyn_cast<AllocaInst>(&inst)) {
+      if (allocaInst->getAllocatedType()->isAggregateType())
+        candidates.push_back(allocaInst);
     }
   }
+  for (auto *alloca : candidates) {
+    if (tryProcessAlloca(alloca))
+      changed = true;
+  }
+  return changed;
 }
 
 // =====================================================================================================================
-// Finds the single "store" instruction storing to this pointer.
-//
-// Returns nullptr if no "store", multiple "store", or partial "store" instructions (store only part
-// of the memory) are found.
-//
-// NOTE: This is conservative in that it returns nullptr if the pointer escapes by being used in anything
-// other than "store" (as the pointer), "load" or "getelementptr" instruction.
+// For a single alloca, try to replace it by a constant global variable.
 //
 // @param allocaInst : The "alloca" instruction to process
-StoreInst *SpirvLowerConstImmediateStore::findSingleStore(AllocaInst *allocaInst) {
-  std::vector<Instruction *> pointers;
-  bool isOuterPointer = true;
-  StoreInst *storeInstFound = nullptr;
-  Instruction *pointer = allocaInst;
-  while (true) {
-    for (auto useIt = pointer->use_begin(), useItEnd = pointer->use_end(); useIt != useItEnd; ++useIt) {
-      auto user = cast<Instruction>(useIt->getUser());
+// @return true if the alloca was replaced
+bool SpirvLowerConstImmediateStore::tryProcessAlloca(AllocaInst *allocaInst) {
+  auto *allocatedTy = allocaInst->getAllocatedType();
+  auto *arrayTy = dyn_cast<ArrayType>(allocatedTy);
+
+  StoreInst *aggregateStore = nullptr;
+  DenseMap<uint64_t, StoreInst *> elementStores;
+
+  SmallVector<Instruction *> toErase;
+  SmallVector<GetElementPtrInst *> geps;
+
+  // Step 1: Determine if the alloca can be converted and find the relevant store(s)
+  SmallVector<std::pair<Value *, std::optional<uint64_t>>> pointers;
+  pointers.emplace_back(allocaInst, 0);
+  do {
+    auto [pointer, index] = pointers.pop_back_val();
+    for (Use &use : pointer->uses()) {
+      auto user = cast<Instruction>(use.getUser());
       if (auto storeInst = dyn_cast<StoreInst>(user)) {
-        if (pointer == storeInst->getValueOperand() || storeInstFound || !isOuterPointer) {
-          // Pointer escapes by being stored, or we have already found a "store"
-          // instruction, or this is a partial "store" instruction.
-          return nullptr;
+        if (pointer == storeInst->getValueOperand() || !index.has_value()) {
```

Comment:
Rather than testing `pointer == storeInst->getValueOperand()` here it would be conceptually cleaner to test for `use` _not_ referring to the address operand of `storeInst`, with something like `use.getOperandNo() != storeInst->getPointerOperandIndex()`.